### PR TITLE
Add digest writer

### PR DESCRIPTION
**Pull request by epapbak**
## PR Content

- Add Kafka consumer and SQL storage handler to retrieve SHA digests from the configured topic and store them for future use
- Grant additional permissions to `digestwriter` on `cluster` table
- Make multiple columns nullable in `cluster` table, as `digestwriter` does not have that information
- Update `Dockerfile` and `docker-compose.yml` files
- Add corresponding UTs

## Secure Coding Practices Checklist GitHub Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Checklist
- [x] Input Validation
- [x] Output Encoding
- [x] Authentication and Password Management
- [x] Session Management
- [x] Access Control
- [x] Cryptographic Practices
- [x] Error Handling and Logging
- [x] Data Protection
- [x] Communication Security
- [x] System Configuration
- [x] Database Security
- [x] File Management
- [x] Memory Management
- [x] General Coding Practices

### CODE REVIEW

**Comment by epapbak** [comment]
> @epapbak Don't you think it should be better to squash the commits (amend) that fix something which is not even merged yet and include this fixed code into the initial feature commit? :) I mean commits like, "clean up comments", "remove method"

I planned on squashing the whole thing when merging, but I just realised that option is disabled. So yes, I will squash some of the commits together to clean it up a little bit.

**Review by jdobes — CHANGES_REQUESTED** [review]
I think this service (service processing data from archives) should be doing something a bit different. It seems writing image SHAs, but it doesn't make much sense - image SHAs are primarily populated by Pyxis Gatherer, populating images here will insert a lot of image SHAs which don't have any known vulnerabilities, so it's not useful to store them.

I've expected it'll rather create/update records in `cluster` table based on the received archive message and records in `cluster_image` for those images, which already exist in DB (previously synced by Pyxis gatherer)

It's possible to insert *all* reported images here and their associations in `cluster_image`, but it'd be a waste of DB space a performance with no use, currently.

File: digestwriter/consumer.go [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+	return NewWithSaramaConfig(DefaultSaramaConfig, storage)
+}
+
+// NewWithSaramaConfig constructs new implementation of Consumer interface with custom sarama config
+func NewWithSaramaConfig(
+	saramaConfig *sarama.Config,
+	storage Storage,
+) (*KafkaConsumer, error) {
+
+	brokerAddress := utils.Getenv("KAFKA_BROKER_ADDRESS", "")
```
[the remark this answers is not in the document]
Comment (epapbak) [reply digestwriter/consumer.go:90]: I'm OK with merging #5 first and going with GetEnv

File: digestwriter/consumer_test.go [hunk 2]
```diff
@@ -0,0 +1,300 @@
+package digestwriter_test
```
[the remark this answers is not in the document]
Comment (epapbak) [reply digestwriter/consumer_test.go:1]: Always =)

File: digestwriter/digestwriter.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	}
+	defer storage.Close()
+
+	consumer, err := startConsumer(storage, logger)
+	if err != nil {
+		return ExitStatusConsumerError, err
+	}
+	defer consumer.Close()
+	return ExitStatusOK, nil
+}
```
Comment (yungbender) [inline digestwriter/digestwriter.go]: newline pls

File: dbadmin/migrations/004_grant_insert_update_image_archive_db_writer.down.sql [hunk 4]
```diff
@@ -0,0 +1,2 @@
+REVOKE INSERT, UPDATE, DELETE ON image IN SCHEMA public FROM archive_db_writer;
+ALTER TABLE image ALTER COLUMN pyxis_id SET NOT NULL;
```
Comment (yungbender) [inline dbadmin/migrations/004_grant_insert_update_image_archive_db_writer.down.sql:2]: newline pls

File: dbadmin/migrations/004_grant_insert_update_image_archive_db_writer.up.sql [hunk 5]
```diff
@@ -0,0 +1,2 @@
+GRANT SELECT, INSERT, UPDATE, DELETE ON image TO archive_db_writer;
+ALTER TABLE image ALTER COLUMN pyxis_id DROP NOT NULL;
```
Comment (yungbender) [inline dbadmin/migrations/004_grant_insert_update_image_archive_db_writer.up.sql:2]: newline pls

File: digestwriter/consumer.go [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+	if msg == nil {
+		logger.Infoln("nil message")
+		return
+	}
+
+	logger.WithFields(logrus.Fields{
+		offsetKey:           msg.Offset,
+		partitionKey:        msg.Partition,
+		topicKey:            msg.Topic,
+		"message_timestamp": msg.Timestamp,
```
Comment (yungbender) [inline digestwriter/consumer.go:245]: why not also move this to the const at the start of the file

File: conf/digestwriter.env [hunk 7]
```diff
@@ -0,0 +1,3 @@
+POSTGRES_USER=archive_db_writer
+POSTGRES_PASSWORD=archive_db_writer_pwd
+LOGGING_LEVEL=DEBUG
```
Comment (yungbender) [inline conf/digestwriter.env:3]: newline pls

File: digestwriter/consumer.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (epapbak) [reply digestwriter/consumer.go]: @jdobes this is fixed. PTAL

File: digestwriter/consumer.go [hunk 8]
```diff
@@ -0,0 +1,351 @@
...
+
+	// key for error message used in structured log messages
+	errorKey = "error"
+
+	// key for duration of message processing used in structured log messages
+	processingDurationKey = "processing_duration"
+)
+
+// ClusterName represents the external cluster UUID contained in the consumed message
+type ClusterName string
```
[the remark this answers is not in the document]
Comment (epapbak) [reply digestwriter/consumer.go:34]: it is used now =)

File: digestwriter/consumer.go [hunk 9]
```diff
@@ -0,0 +1,351 @@
...
+	// Step #1: parse the incoming message
+	message, err := parseMessage(msg.Value)
+	if err != nil {
+		/* ParseIncomingMessageError.Inc() */
+		return err
+	}
+
+	/* ParsedIncomingMessage.Inc() */
+
+	logger.Infoln("Parsed incoming message successfully")
```
[the remark this answers is not in the document]
Comment (epapbak) [reply digestwriter/consumer.go]: I changed most of the logging to debug. PTAL.

File: digestwriter/consumer.go [hunk 10]
```diff
@@ -0,0 +1,351 @@
...
+// GetNumberOfErrorsConsumingMessages returns number of errors during consuming messages
+// since creating KafkaConsumer obj
+func (consumer *KafkaConsumer) GetNumberOfErrorsConsumingMessages() uint64 {
+	return consumer.numberOfErrorsConsumingMessages
+}
+
+// handleMessage handles the message and does all logging, metrics, etc
+func (consumer *KafkaConsumer) handleMessage(msg *sarama.ConsumerMessage) {
+	if msg == nil {
+		logger.Infoln("nil message")
```
[the remark this answers is not in the document]
Comment (epapbak) [reply digestwriter/consumer.go]: I changed most of the logging to debug. PTAL.

File: digestwriter/consumer.go [hunk 11]
```diff
@@ -0,0 +1,373 @@
...
+// useful for testing
+var DefaultSaramaConfig *sarama.Config
+
+// NewConsumer constructs new implementation of Consumer interface
+func NewConsumer(storage Storage, logger *logrus.Logger) (*KafkaConsumer, error) {
+	return NewWithSaramaConfig(DefaultSaramaConfig, storage, logger)
+}
+
+// NewWithSaramaConfig constructs new implementation of Consumer interface with custom sarama config
+func NewWithSaramaConfig(
```
[the remark this answers is not in the document]
Comment (epapbak) [reply digestwriter/consumer.go:114]: I'd prefer to do that in another commit,  if you don't mind.

File: dbadmin/migrations/006_alter_cluster_table_for_archive_db_writer.up.sql [hunk 12]
```diff
@@ -0,0 +1,7 @@
+GRANT SELECT, INSERT, UPDATE, DELETE ON cluster TO archive_db_writer;
+ALTER TABLE cluster ALTER COLUMN status DROP NOT NULL;
+ALTER TABLE cluster ALTER COLUMN version DROP NOT NULL;
+ALTER TABLE cluster ALTER COLUMN cve_cache_critical DROP NOT NULL;
```
Comment (jdobes) [inline dbadmin/migrations/006_alter_cluster_table_for_archive_db_writer.up.sql:4]: not necessary to drop `NOT NULL` for these `cve_cache_*`, they have `DEFAULT 0`, so they will always have value

File: digestwriter/storage.go [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+	}
+
+	logger.WithFields(logrus.Fields{
+		"clusterID": clusterInfoData.ID,
+	}).Infoln("updated cluster table successfully")
+
+	if err = storage.linkDigestsToCluster(tx, clusterInfoData.ID, digests); err != nil {
+		logger.WithFields(logrus.Fields{
+			"clusterID": clusterInfoData.ID,
+		}).Infoln("we died here!")
```
Comment (jdobes) [inline digestwriter/storage.go:169]: Errorln?

**Review by jdobes — CHANGES_REQUESTED** [review]
Small things only. Otherwise looks good to me and I'd merge it to not block other changes. We need the kafka and data coming in to test it properly.

**Comment by jdobes** [comment]
And sorry for another conflict in go.mod.

File: dbadmin/migrations/006_alter_cluster_table_for_archive_db_writer.up.sql (hunk 12 above)
Comment (epapbak) [reply dbadmin/migrations/006_alter_cluster_table_for_archive_db_writer.up.sql:4]: true

**Comment by epapbak** [comment]
> Small things only. Otherwise looks good to me and I'd merge it to not block other changes. We need the kafka and data coming in to test it properly.

I'm working on that on another PR. See https://github.com/epapbak/vuln4shift-backend/tree/digestwriter_localsetup

**Review by jdobes — APPROVED** [review]

**Pull request merged**